### PR TITLE
add oidc config.

### PR DESCRIPTION
**Pull request by Alice1319**
fixes eclipse-xpanse/xpanse#572
fixes eclipse-xpanse/xpanse#574
fixes eclipse-xpanse/xpanse#581
fixes eclipse-xpanse/xpanse#593

image

image

image

image

image

### CODE REVIEW

File: src/components/config/OidcConfig.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * SPDX-FileCopyrightText: Huawei Inc.
+ */
+
+import { AuthProvider, User } from 'oidc-react';
```
Comment (swaroopar) [inline src/components/config/OidcConfig.tsx:6]: oidc-react package is missing in the package.json and package-lock.json

File: src/components/config/config.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * SPDX-FileCopyrightText: Huawei Inc.
+ */
+export const authority: string = 'https://iam.xpanse.site';
```
Comment (swaroopar) [inline src/components/config/config.ts:5]: all environment specific config must move to .env files.

File: src/components/config/OidcConfig.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            if (response === null) {
+                return;
+            }
+            localStorage.setItem(configResponseKey, response.toStorageString());
+            localStorage.setItem(currentUserKey, response.profile.name === undefined ? '' : response.profile.name);
+            const roles: unknown = response.profile['urn:zitadel:iam:org:project:roles'];
+            if (Array.isArray(roles)) {
+                // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+                const objectKeys = Object.keys(roles[0]);
+                localStorage.setItem(isAuthenticatedKey, 'true');
```
Comment (swaroopar) [inline src/components/config/OidcConfig.tsx:37]: we must avoid setting these values. 
Instead use the 'useAuth' hook to get the values.

**Comment by swaroopar** [comment]
@Alice1319 
It looks like the oidc-react library does not satisfy all our requirements. We must switch to another one - https://github.com/AxaFrance/react-oidc

1. I have used your branch and created another branch just to show you how it must be done. You can refer to my changes here - https://github.com/eclipse-xpanse/xpanse-ui/tree/feature/use-react-oidc
2. Important is -> complete app must be projected like this with a context. https://github.com/eclipse-xpanse/xpanse-ui/blob/feature/use-react-oidc/src/App.tsx#L35
3. additionally all routes must be protected like this - https://github.com/eclipse-xpanse/xpanse-ui/blob/feature/use-react-oidc/src/App.tsx#L40
4. Then within the "Protected" component, we only verify the roles. https://github.com/eclipse-xpanse/xpanse-ui/blob/feature/use-react-oidc/src/components/protectedRoutes/ProtectedRoute.tsx#L39 -> this method is not fully implemented. I see one small problem here that the userInfo is null for few seconds and its only available after few seconds. Please see if you find any workaround to handle this. We need some kind of wait there. 
5. Important is - nothing should be stored and read from local storage. Everything should be accessed from hooks only. 

Please copy whatever changes you find right from my branch to yours.

**Comment by iskey** [comment]
As we are taking the csp as the default role when login, this may puzzle someone, would it be better to record the last selection?

File: src/components/config/OidcConfig.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (swaroopar) [inline src/components/config/OidcConfig.tsx:1]: rename the folder as oidc.

**Review by swaroopar — CHANGES_REQUESTED** [review]

File: src/components/config/OidcConfig.tsx (hunk 1 above)
Comment (Alice1319) [reply src/components/config/OidcConfig.tsx:6]: fixed.

File: src/components/config/OidcConfig.tsx (hunk 4 above)
Comment (Alice1319) [reply src/components/config/OidcConfig.tsx]: fixed.

File: src/App.tsx [hunk 5]
```diff
@@ -27,84 +26,120 @@ import OrderSubmitPage from './components/content/order/OrderSubmit';
...
-                    <Protected allowedRole={'user'}>
-                        <Credential />
-                    </Protected>
-                }
-            />
-            <Route path='*' element={<LoginScreen />} />
-        </Routes>
+        <OidcProvider
+            configuration={OidcConfig}
+            onEvent={(configurationName, eventName, data) => {
```
Comment (swaroopar) [inline src/App.tsx:36]: we can remove this onEvent method. It was only for testing.

File: src/components/oidc/OidcConfig.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+export const onEvent = (configurationName: string, eventName: string, data: object) => {
+    console.log(`oidc:${configurationName}:${eventName}`, data);
+};
+
+export const OidcConfig: OidcConfiguration = {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    authority: process.env.REACT_APP_ZITADEL_AUTHORITY_NAME!,
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    client_id: process.env.REACT_APP_ZITADEL_CLIENT_ID!,
+    redirect_uri: window.location.origin + '/authentication/callback',
```
Comment (swaroopar) [inline src/components/oidc/OidcConfig.tsx:17]: /authentication/callback and /authentication/silent-callback also must be moved to .env file

**Comment by swaroopar** [comment]
@Alice1319 As we discussed, please remove all "localstorage" calls. Use the hooks form react-oidc lib. 
To address the point raised by Yichao, we can store the last selected role in the sessionstore (not local storage) and compare and use the same if the fresh login also returns the the stored role as one of the available roles of the user.

**Comment by Alice1319** [comment]
> As we are taking the csp as the default role when login, this may puzzle someone, would it be better to record the last selection?

reference  to eclipse-xpanse/xpanse/#593

**Comment by Alice1319** [comment]
> @Alice1319 As we discussed, please remove all "localstorage" calls. Use the hooks form react-oidc lib. To address the point raised by Yichao, we can store the last selected role in the sessionstore (not local storage) and compare and use the same if the fresh login also returns the the stored role as one of the available roles of the user.

reference  to eclipse-xpanse/xpanse#593

File: src/components/content/credential/Credential.tsx [hunk 7]
```diff
@@ -126,7 +126,7 @@ function Credential(): JSX.Element {
 
     const updateCredential = (abstractCredentialInfo: AbstractCredentialInfo) => {
         const credentialVariables: CredentialVariables = abstractCredentialInfo;
-        const userName: string | null = localStorage.getItem(usernameKey);
+        const userName: string | null = sessionStorage.getItem(userRoleKey);
```
Comment (swaroopar) [inline src/components/content/credential/Credential.tsx:129]: here also we must get the username from token. Not the  role name. 
Role in session storage must be used only for setting the previous selected role. For nothing else.

File: src/components/content/credential/Credential.tsx [hunk 8]
```diff
@@ -157,7 +157,7 @@ function Credential(): JSX.Element {
     };
 
     function getCredentials(): void {
-        const userName: string | null = localStorage.getItem(usernameKey);
+        const userName: string | null = sessionStorage.getItem(userRoleKey);
```
Comment (swaroopar) [inline src/components/content/credential/Credential.tsx:160]: same here. username must be taken from token.

File: src/components/content/credential/UpdateCredential.tsx [hunk 9]
```diff
@@ -44,7 +44,7 @@ function UpdateCredential({
 
     const submit = (createCredential: CreateCredential) => {
         if (!isContainsEmpty(createCredential.variables)) {
-            const userName: string | null = localStorage.getItem(usernameKey);
+            const userName: string | null = sessionStorage.getItem(userRoleKey);
```
Comment (swaroopar) [inline src/components/content/credential/UpdateCredential.tsx:47]: same here. username must be taken from token.

File: src/components/content/monitor/Monitor.tsx [hunk 10]
```diff
@@ -30,7 +30,7 @@ function Monitor(): JSX.Element {
     >([{ value: '', label: '', serviceName: '', id: '' }]);
 
     useEffect(() => {
-        const userName: string | null = localStorage.getItem(usernameKey);
+        const userName: string | null = sessionStorage.getItem(userRoleKey);
```
Comment (swaroopar) [inline src/components/content/monitor/Monitor.tsx:33]: same here. username must be taken from token.

File: src/components/content/order/OrderSubmit.tsx [hunk 11]
```diff
@@ -120,7 +120,7 @@ function OrderSubmit(props: OrderSubmitProps): JSX.Element {
             )
         );
         // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        ServiceService.getDeployedServiceDetailsById(uuid, localStorage.getItem(usernameKey)!)
+        ServiceService.getDeployedServiceDetailsById(uuid, sessionStorage.getItem(userRoleKey)!)
```
Comment (swaroopar) [inline src/components/content/order/OrderSubmit.tsx:123]: same here. username must be taken from token.

File: src/components/content/order/OrderSubmit.tsx [hunk 12]
```diff
@@ -172,7 +172,7 @@ function OrderSubmit(props: OrderSubmitProps): JSX.Element {
             version: props.version,
             customerServiceName: customerServiceName,
             // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-            userName: localStorage.getItem(usernameKey)!,
+            userName: sessionStorage.getItem(userRoleKey)!,
```
Comment (swaroopar) [inline src/components/content/order/OrderSubmit.tsx:175]: same here. username must be taken from token.

File: src/components/content/order/ServiceList.tsx [hunk 13]
```diff
@@ -179,7 +179,7 @@ function ServiceList(): JSX.Element {
             'Destroying, Please wait... [' + Math.ceil((new Date().getTime() - date.getTime()) / 1000).toString() + 's]'
         );
         // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        ServiceService.getDeployedServiceDetailsById(uuid, localStorage.getItem(usernameKey)!)
+        ServiceService.getDeployedServiceDetailsById(uuid, sessionStorage.getItem(userRoleKey)!)
```
Comment (swaroopar) [inline src/components/content/order/ServiceList.tsx:182]: same here. username must be taken from token.

File: src/components/content/order/ServiceList.tsx [hunk 14]
```diff
@@ -232,7 +232,7 @@ function ServiceList(): JSX.Element {
 
     function getDeployedProperties(id: string): void {
         // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        ServiceService.getDeployedServiceDetailsById(id, localStorage.getItem(usernameKey)!)
+        ServiceService.getDeployedServiceDetailsById(id, sessionStorage.getItem(userRoleKey)!)
```
Comment (swaroopar) [inline src/components/content/order/ServiceList.tsx:235]: same here. username must be taken from token.

File: src/components/content/order/ServiceList.tsx [hunk 15]
```diff
@@ -350,7 +350,7 @@ function ServiceList(): JSX.Element {
 
     function getServices(): void {
         setServicesLoadingError(<></>);
-        const userName: string | null = localStorage.getItem(usernameKey);
+        const userName: string | null = sessionStorage.getItem(userRoleKey);
```
Comment (swaroopar) [inline src/components/content/order/ServiceList.tsx:353]: same here. username must be taken from token.

File: src/components/content/order/formElements/CommonTypes.ts [hunk 16]
```diff
@@ -178,7 +178,7 @@ export const getCreateRequest = (props: OrderSubmitProps, customerServiceName: s
         version: props.version,
         customerServiceName: customerServiceName,
         // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        userName: localStorage.getItem(usernameKey)!,
+        userName: sessionStorage.getItem(userRoleKey)!,
```
Comment (swaroopar) [inline src/components/content/order/formElements/CommonTypes.ts:181]: same here. username must be taken from token.

File: src/components/content/order/migrate/MigrateService.tsx [hunk 17]
```diff
@@ -160,7 +160,7 @@ export const MigrateService = ({
             )
         );
         // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        ServiceService.getDeployedServiceDetailsById(uuid, localStorage.getItem(usernameKey)!)
+        ServiceService.getDeployedServiceDetailsById(uuid, sessionStorage.getItem(userRoleKey)!)
```
Comment (swaroopar) [inline src/components/content/order/migrate/MigrateService.tsx:163]: same here. username must be taken from token.

File: src/components/content/order/migrate/MigrateService.tsx [hunk 18]
```diff
@@ -258,7 +258,7 @@ export const MigrateService = ({
...
```
Comment (swaroopar) [inline src/components/content/order/migrate/MigrateService.tsx:261]: same here. username must be taken from token.

**Review by swaroopar — CHANGES_REQUESTED** [review]
comments added.

**Comment by WangLiNaruto** [comment]
looks good

**Review by swaroopar — APPROVED** [review]
LGTM

**Pull request merged**